### PR TITLE
Add container remote APIs

### PR DESCRIPTION
**Pull request by newswangerd**
## New APIs
`/api/automation-hub/_ui/v1/execution-environments/remotes/`

Allows creating updating and viewing container remotes.

Creating a new remote with this endpoint will automatically initialize a container repository and distribution to go alongside the remote. It requires a registry to be selected when creating a new remote.  The selected registry is linked to the new remote and the connection information for the registry is copied into the remote object.

### Sample response

```

{
    "meta": {
        "count": 1
    },
    "links": {
        "first": "/api/automation-hub/_ui/v1/execution-environments/remotes/?limit=10&offset=0",
        "previous": null,
        "next": null,
        "last": "/api/automation-hub/_ui/v1/execution-environments/remotes/?limit=10&offset=0"
    },
    "data": [
        {
            "pulp_id": "ba95ac4e-4cd3-4a2e-b531-b4a4d669a2ef",
            "name": "python",
            "upstream_name": "newswangerd/ubuntu",
            "registry": "e538f3ae-6e37-4d77-9644-431fe9d41dde",
            "last_sync_task": {
                "task_id": "950dc0aa-f4c2-4826-8925-452df33ecbb6",
                "state": "completed",
                "started_at": "2021-09-02T19:50:12.011078Z",
                "finished_at": "2021-09-02T19:50:15.658338Z",
                "error": null
            }
        }
    ]
}
```

## Changes
- The data models for last sync task for collection and container repositories have been removed and their functionality has been replaced by querying the pulp tasking database for reserved resources
- Added remote information to the container repository endpoint
- Fixes an issue with the registry remote that was causing the DRF forms to break on the pulp repository APIs.

## Todo
- [x] Write tests
- [x] Add access policy

### CODE REVIEW

**Comment by himdel** [comment]
Looks like I can successfully create a remote :+1: ...

```
ExecutionEnvironmentRemoteAPI.create({
    "name": "string!@#$%^&*()_-=+/?.,<>:;",
    "upstream_name": "string!@#$%^&*()_-=+/?.,<>:;",
    "registry": "27ade43c-a0ab-4963-a73b-c4cb58ef16a9",
})
```

but maybe we should do some valudation on name & upstream_name

And after that, the `_ui/v1/execution-environments/repositories/` (Container Registry list) starts failing..

```
api_1          | Internal Server Error: /api/automation-hub/_ui/v1/execution-environments/repositories/
api_1          | Traceback (most recent call last):
api_1          |   File "/venv/lib64/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
api_1          |     response = get_response(request)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
api_1          |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
api_1          |     return view_func(*args, **kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/viewsets.py", line 125, in view
api_1          |     return self.dispatch(request, *args, **kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 509, in dispatch
api_1          |     response = self.handle_exception(exc)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 469, in handle_exception
api_1          |     self.raise_uncaught_exception(exc)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
api_1          |     raise exc
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/views.py", line 506, in dispatch
api_1          |     response = handler(request, *args, **kwargs)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/mixins.py", line 43, in list
api_1          |     return self.get_paginated_response(serializer.data)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/serializers.py", line 745, in data
api_1          |     ret = super().data
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/serializers.py", line 246, in data
api_1          |     self._data = self.to_representation(self.instance)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/serializers.py", line 663, in to_representation
api_1          |     return [
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/serializers.py", line 664, in <listcomp>
api_1          |     self.child.to_representation(item) for item in iterable
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/serializers.py", line 515, in to_representation
api_1          |     ret[field.field_name] = field.to_representation(attribute)
api_1          |   File "/venv/lib64/python3.8/site-packages/rest_framework/fields.py", line 1882, in to_representation
api_1          |     return method(value)
api_1          |   File "/src/galaxy_ng/galaxy_ng/app/api/ui/serializers/execution_environment.py", line 109, in get_pulp
api_1          |     "last_sync_task": _get_last_sync_task(repo),
api_1          |   File "/src/galaxy_ng/galaxy_ng/app/api/ui/serializers/execution_environment.py", line 124, in _get_last_sync_task
api_1          |     sync_task = models.container.ContainerSyncTask.objects.filter(repository=repo).first()
api_1          | AttributeError: module 'galaxy_ng.app.models.container' has no attribute 'ContainerSyncTask'
api_1          | pulp [None]: django.request:ERROR: Internal Server Error: /api/automation-hub/_ui/v1/execution-environments/repositories/
```

**Comment by himdel** [comment]
(hacked around the issue with `sync_task=None`...)

I can successfully edit descriiption and readme of remotes with the existing form, they do appear in the repositories list, and I can tell them apart using `.pulp.repository.pulp_type: "container.container-push"` (and `remote: null`) vs `container.container` and non-null `remote` :+1: 

(my todo: sync, delete, include_tags/exclude_tags)

**Comment by himdel** [comment]
I can't get sync to work...

GET /api/automation-hub/_ui/v1/execution-environments/remotes/cb92a22c-08f8-4b68-bffa-26dfdfe535b3/
returns a remote

POST /api/automation-hub/_ui/v1/execution-environments/remotes/cb92a22c-08f8-4b68-bffa-26dfdfe535b3/sync/
returns 404

it doesn't seem to be mentioned in URLs

**Comment by newswangerd** [comment]
@himdel you need to run the migrations. I removed the container sync task database table

**Comment by himdel** [comment]
@newswangerd you did, but this PR was started before #898 was merged, and it adds a `ContainerSyncTask` use that you haven't removed .. at least I think that's what I'm seeing :) Can you try rebasing and looking for `ContainerSyncTask` again?

File: galaxy_ng/app/api/v3/serializers/sync.py [hunk 1]
```diff
@@ -29,32 +28,14 @@ class Meta:
...
-            "started_at": sync_task.task.started_at,
-            "finished_at": sync_task.task.finished_at,
-            "error": sync_task.task.error
-        }
-
-
-class AnsibleRepositorySerializer(LastSyncTaskMixin, serializers.ModelSerializer):
+class AnsibleRepositorySerializer(serializers.ModelSerializer):
     distributions = serializers.SerializerMethodField()
     last_sync_task = serializers.SerializerMethodField()
```
Comment (rochacbruno) [inline galaxy_ng/app/api/v3/serializers/sync.py:33]: remove this line?

Comment (rochacbruno) [reply galaxy_ng/app/api/v3/serializers/sync.py:33]: this field is redefined below

File: galaxy_ng/app/api/ui/serializers/execution_environment.py [hunk 2]
```diff
@@ -88,6 +94,9 @@ def get_updated(self, distro):
 
     def get_pulp(self, distro):
         repo = distro.repository
+        remote = None
+        if(repo.remote):
```
Comment (rochacbruno) [inline galaxy_ng/app/api/ui/serializers/execution_environment.py:98]: unnecessary parenthesis 

```suggestion
        if repo.remote:
```

Comment (newswangerd) [reply galaxy_ng/app/api/ui/serializers/execution_environment.py:98]: woops. Javascript habits die hard.

Comment (newswangerd) [reply galaxy_ng/app/api/ui/serializers/execution_environment.py:98]: I'm surprised flake8 doesn't catch stuff like that. It cares about the most pedantic things like having a new line character at the end of each file.

**Review by rochacbruno — APPROVED** [review]

**Pull request merged**